webapp/src/store/ProjectStore.ts

```diff
@@ -1,28 +1,34 @@
 import {
+  IMessageAdapter,
   ITranslationAdapter,
+  MessageData,
   MessageMap,
   TranslationMap,
 } from '@/utils/adapters';
 import { LanguageNotFound, MessageNotFound } from '@/errors';
-
-type StoreData = {
-  languages: TranslationMap;
-};
+import { StoreData } from './types';
+import mergeStoreData from './mergeStoreData';
 
 export class ProjectStore {
   private data: StoreData;
   private translationAdapter: ITranslationAdapter;
+  private messageAdapter: IMessageAdapter;
 
-  constructor(translationAdapter: ITranslationAdapter) {
+  constructor(
+    messageAdapter: IMessageAdapter,
+    translationAdapter: ITranslationAdapter,
+  ) {
     this.data = {
       languages: {},
+      messages: [],
```

Comment:
Are you suggesting we replace `this.data` with just `this.languages`? We need the full `StoreData` to be able to pass it to `mergeStoreData()`, but I guess we could create an "ephemeral" `StoreData` object in `refresh()` in order to be able to pass it to `mergeStoreData()`. Do you think that would be better?

Personally, I think it's easier to think of `data` as a [data clump](https://refactoring.guru/smells/data-clumps) that always goes together, and hence deserves it's own object type.

Reply:
Yes, I am thinking about removing `messages` from `data`. The function `mergeStoreData` needs access to messages corresponding to the `fromRepo.languages`, but it does not need messages corresponding to `inMemory.languages`.

I like that the current implementation illustrates the idea of a complete state, parsed from the filesystem, and the merging of that complete state. The code ended up entirely ignoring `data.messages` so this neat idea turned out to be not exactly what Lyra needs, at least not right now. I don't like code that is effectively unused and ideally I want the implementation of Lyra to not contain concepts and code paths that serve no purpose. I don't find potential future purposes very attractive motivators for unused code either. The future is too uncertain and too distant.

Anyways, it's no big deal. I'll explore this later but the code is fine as is. Even if it wasn't, the benefits of this change set by far outweigh this itch.

Reply:
Correct me if I'm wrong, but in my mind the purpose of having `messages` in memory is so that we do not have to read them from file every time they're needed. IMO, that is not just a hypothetical, but a very reasonable expectation.

We did decide not to build such caching in this PR (in fact we went in the other direction and removed caching that already existed, for translations) but in my mind I considered that a short-term sacrifice in order for our current task (merging states) to be simpler and more focused.

I think it makes sense to re-implement caching quite soon. Even if it's only a minor performance problem right now, it is also a locking problem (as you've pointed out in other comments). The effects on performance will also get worse as reading becomes more complex, which it will if we want to implement timestamps on messages as that will require a much slower traversal of git history.

So all in all I think there are two main reasons for why `messages` belong here, in `StoreData` and why `translations` shouldn't be stored separate:

1. It makes sense conceptually to treat messages and translations as a data clump, that belongs together in an object (`DataStore`)
2. We strongly believe (or at least I do) that we will need it soon to (A) prevent race conditions and (B) make the code history-aware

While I generally agree that premature design can be problematic, I think point 2A is becoming urgent and point 2B is not too far into the future, and even that wasn't the case I think point 1 still stands on it's own.

But with that I've made my case and I'm happy for you to do whatever you think is necessary to feel confident about the code!